icu4c/source/test/intltest/listformattertest.cpp

```diff
@@ -161,6 +214,154 @@ void ListFormatterTest::TestEnglishGB() {
+        data, 3, expected, tupleCount, appendTo,
+        "TestFieldPositionIteratorWith3ItemsAndDataBefore");
+}
+
+void ListFormatterTest::TestFieldPositionIteratorWith3Items() {
+    //  0         1
+    //  012345678901234
+    // "a, bbb, and cc"
+    UnicodeString data[3] = {"a", "bbb", "cc"};
+    int32_t expected[] = {
+        ULISTFMT_ELEMENT_FIELD, 0, 1,
+        ULISTFMT_LITERAL_FIELD, 1, 3,
+        ULISTFMT_ELEMENT_FIELD, 3, 6,
+        ULISTFMT_LITERAL_FIELD, 6, 12,
+        ULISTFMT_ELEMENT_FIELD, 12, 14
+    };
+    int32_t tupleCount = sizeof(expected)/(3 * sizeof(*expected));
+    UnicodeString appendTo;
+    RunTestFieldPositionIteratorWithNItems(
+        data, 3, expected, tupleCount, appendTo,
+        "TestFieldPositionIteratorWith3Items");
+}
+
+void ListFormatterTest::TestFieldPositionIteratorWith2ItemsAndDataBefore() {
+    //  0         1
+    //  0123456789012345
+    // "Foo: bbb and cc"
+    UnicodeString data[2] = {"bbb", "cc"};
+    int32_t expected[] = {
+        ULISTFMT_ELEMENT_FIELD, 5, 8,
+        ULISTFMT_LITERAL_FIELD, 8, 13,
+        ULISTFMT_ELEMENT_FIELD, 13, 15
+    };
+    int32_t tupleCount = sizeof(expected)/(3 * sizeof(*expected));
+    UnicodeString appendTo("Foo: ");
+    RunTestFieldPositionIteratorWithNItems(
+        data, 2, expected, tupleCount, appendTo,
+        "TestFieldPositionIteratorWith2ItemsAndDataBefore");
+}
+
+void ListFormatterTest::TestFieldPositionIteratorWith2Items() {
+    //  0         1
+    //  01234567890
+    // "bbb and cc"
+    UnicodeString data[2] = {"bbb", "cc"};
+    int32_t expected[] = {
+        ULISTFMT_ELEMENT_FIELD, 0, 3,
+        ULISTFMT_LITERAL_FIELD, 3, 8,
+        ULISTFMT_ELEMENT_FIELD, 8, 10
+    };
+    int32_t tupleCount = sizeof(expected)/(3 * sizeof(*expected));
+    UnicodeString appendTo;
+    RunTestFieldPositionIteratorWithNItems(
+        data, 2, expected, tupleCount, appendTo,
+        "TestFieldPositionIteratorWith2Items");
+}
+
+void ListFormatterTest::TestFieldPositionIteratorWith1ItemAndDataBefore() {
+    //  012345678
+    // "Hello cc"
+    UnicodeString data[1] = {"cc"};
+    int32_t expected[] = {
+        ULISTFMT_ELEMENT_FIELD, 6, 8
+    };
+    int32_t tupleCount = sizeof(expected)/(3 * sizeof(*expected));
+    UnicodeString appendTo("Hello ");
+    RunTestFieldPositionIteratorWithNItems(
+        data, 1, expected, tupleCount, appendTo,
+        "TestFieldPositionIteratorWith1ItemAndDataBefore");
+}
+
+void ListFormatterTest::TestFieldPositionIteratorWith1Item() {
+    //  012
+    // "cc"
+    UnicodeString data[1] = {"cc"};
+    int32_t expected[] = {
+        ULISTFMT_ELEMENT_FIELD, 0, 2
+    };
+    int32_t tupleCount = sizeof(expected)/(3 * sizeof(*expected));
+    UnicodeString appendTo;
+    RunTestFieldPositionIteratorWithNItems(
+        data, 1, expected, tupleCount, appendTo,
+        "TestFieldPositionIteratorWith1Item");
+}
+
 // Tests resource loading and inheritance when region sublocale
```

Comment:
Can you add a test for positions with some locale whose pattern format has a prefix that it adds all the time, such that `offsetFirst != 0`?  Because the English pattern isn't particularly interesting to test that part of the business logic.

Reply:
looks like the only locale which has pattern that not started from {0} is ur_IN
egrep "[^a-zA-Z0-9](start|middle|end|2){" third_party/icu/source/data/locales/*|egrep -v "\"{0"
third_party/icu/source/data/locales/ur_IN.txt:            2{"{1}، {0}"}
third_party/icu/source/data/locales/ur_IN.txt:            end{"{1}، {0}"}
third_party/icu/source/data/locales/ur_IN.txt:            2{"{1} {0}"}
third_party/icu/source/data/locales/ur_IN.txt:            end{"{1} {0}"}
third_party/icu/source/data/locales/ur_IN.txt:            middle{"{1} {0}"}
third_party/icu/source/data/locales/ur_IN.txt:            start{"{1} {0}"}
third_party/icu/source/data/locales/ur_IN.txt:            2{"{1}، {0}"}
third_party/icu/source/data/locales/ur_IN.txt:            end{"{1}، {0}"}

Reply:
done

Reply:
I don't see the new test case.  Did you forget to upload it?

Reply:
see https://github.com/unicode-org/icu/pull/106/files#diff-0ffa21a36a6c9102daf8d2dbbe9aedabR337 and
https://github.com/unicode-org/icu/pull/106/files#diff-0ffa21a36a6c9102daf8d2dbbe9aedabR393

Reply:
Ok, I see the "ur" case.  What I was looking for was a case that tested for the correctness of the `prefixLength` logic, but it seems that there is no such locale that exercises that case.